libdnf/rpm/package_query.cpp

```diff
@@ -2456,6 +2456,17 @@ void PackageQuery::filter_recent(const time_t timestamp) {
     }
 }
 
+void PackageQuery::filter_userinstalled() {
+    filter_installed();
+    for (const auto & pkg : *this) {
+        auto reason = pkg.get_reason();
+        if (reason == libdnf::transaction::TransactionItemReason::WEAK_DEPENDENCY ||
+            reason == libdnf::transaction::TransactionItemReason::DEPENDENCY) {
+            p_impl->remove_unsafe(pkg.get_id().id);
```

Comment:
This is a minor note but interesting - p_impl->remove_unsafe() is in most of cases slower then creating empty PackageSet, add passing candidate and do `and` operator. The trick is the number operations. If I have 1000 packages and 100 is user installed then it requires 900 time operation on bitmap. The other approach requires only 100 modifications and then and operation is very cheap. The first version of DNF5 queries used remove_unsafe() approach and affected performance was significantly affected.